src/com/facebook/buck/jvm/kotlin/KotlinBuckConfig.java

```diff
@@ -132,6 +135,49 @@ Path getPathToCompilerJar() {
   }
 
   /**
+   * Get the path to the Kotlin annotation processing jar.
+   *
+   * @return the Kotlin annotation processing jar path
+   */
+  Path getPathToAP() {
+    Path compiler = getKotlinHome().resolve("kotlin-annotation-processing.jar");
+    if (Files.isRegularFile(compiler)) {
+      return compiler.normalize();
+    }
+
+    compiler = getKotlinHome().resolve(Paths.get("lib", "kotlin-annotation-processing.jar"));
+    if (Files.isRegularFile(compiler)) {
+      return compiler.normalize();
+    }
+
+    compiler =
+        getKotlinHome().resolve(Paths.get("libexec", "lib", "kotlin-annotation-processing.jar"));
+    if (Files.isRegularFile(compiler)) {
+      return compiler.normalize();
+    }
+
+    throw new HumanReadableException(
+        "Could not resolve kotlin annotation processing JAR location (kotlin home:"
+            + getKotlinHome()
+            + ").");
+  }
+
+  /**
+   * Get the path to the java tools jar.
+   *
+   * @return the java tools jar path
+   */
+  Path getPathToTools() {
+    Path compiler = getJavaHome().resolve("lib/tools.jar");
```

Comment:
This is no longer present in jdk 9. Can we not need this or atleast let users specify a location for this jar?

Reply:
Currently it is required when building with JDK 7 or JDK 8. 

According to https://docs.oracle.com/javase/9/migrate/toc.htm#JSMIG-GUID-A78CC891-701D-4549-AA4E-B8DD90228B4B the tools jar may be instead replaced by a "more efficient format in implementation-specific files in the lib directory." so we could potentially depend on whatever in tools.jar kapt requires when building with JDK 9.

I can change the implementation for now to allow users to specify the location. But I'll keep digging to see what kapt in Gradle will do when building with JDK 9 if it is supported yet.

Reply:
It looks like this path is used to initialize a `ClassLoader` that is used to load the compiler? If that's the only usage, `ToolProvider.getSystemToolClassLoader` will get you a `ClassLoader` that is equivalent to including `tools.jar`. You can use that as the parent `ClassLoader`.

Note that `ToolProvider` is not actually thread safe (yay!) so we need to ensure all our usages of it are synchronized. Right now the only usage is in `JdkProvidedInMemoryJavac`; you'll need to factor out some new synchronized wrapper class to use both here and there.

That will allow you to get rid of the scary `java_home` finding logic too right?

Reply:
I've found a mention, and explains where tools.jar is used on kapt [here](https://github.com/JetBrains/kotlin/blob/0b37c9e83cd09008db5908fd47583cd62e9fc17b/plugins/kapt3/kapt3-compiler/src/org/jetbrains/kotlin/kapt3/Kapt3Plugin.kt#L210)

Reply:
That reference doesn't contradict what I'm saying...it's expecting to be able to use its default `ClassLoader` to load certain classes that live in `tools.jar` on Java 8-. Controlling the `ClassLoader` as I describe above should work for that without having to directly reference `tools.jar`.

Reply:
I really don't know, the compile expects `tools.jar` as a plugin, and use it internally. I'll have to see if there is a way to provide this ClassLoader directly.

Reply:
@thalescm The clue is in https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/kotlin/JarBackedReflectedKotlinc.java#L160

This is where the `Kotlinc` is loaded by creating a new classloader with the kotlin jars passed in via buckconfig. you should change the parent classloader from null to `ToolProvider.getSystemToolClassLoader`

This way, classloader hierarchy will set setup such that you dont have to add the tools jar again when running the `KotlincStep`